src/lambertw.jl

```diff
@@ -0,0 +1,386 @@
+the complex plane.
+
+```jldoctest
+julia> lambertw(-1/e,-1)
+-1.0
+
+julia> lambertw(-1/e,0)
+-1.0
+
+julia> lambertw(0,0)
+0.0
+
+julia> lambertw(0,-1)
+-Inf
+
+julia> lambertw(Complex(-10.0,3.0), 4)
+-0.9274337508660128 + 26.37693445371142im
+```
+
+!!! note
+    The constant `LAMBERTW_USE_NAN` at the top of the source file controls whether arguments
+    outside the domain throw `DomainError` or return `NaN`. The default is `DomainError`.
+"""
+function lambertw(x::Real, k::Integer)
+    k == 0 && return lambertwk0(x)
+    k == -1 && return _lambertwkm1(x)
+    @baddomain(k)  # more informative message like below ?
+#    error("lambertw: real x must have k == 0 or k == -1")
+end
+
+function lambertw(x::Union{Integer,Rational}, k::Integer)
+    if k == 0
+        x == 0 && return float(zero(x))
+        x == 1 && return convert(typeof(float(x)),omega) # must be more efficient way
+    end
+    lambertw(float(x),k)
+end
+
+### Complex z ###
+
+# choose initial value inside correct branch for root finding
+function lambertw(z::Complex{T}, k::Integer) where T<:Real
+    one_t = one(T)
+    local w::Complex{T}
+    pointseven = 7//10
+    if abs(z) <= one_t/convert(T,euler)
+        if z == 0
+            k == 0 && return z
+            return complex(-convert(T,Inf),zero(T))
+        end
+        if k == 0
+            w = z
+        elseif k == -1 && imag(z) == 0 && real(z) < 0
+            w = complex(log(-real(z)),1//10^7) # need offset for z ≈ -1/e.
+        else
+            w = log(z)
+            k != 0 ? w += complex(0,k * 2 * pi) : nothing
+        end
+    elseif k == 0 && imag(z) <= pointseven && abs(z) <= pointseven
+        w = abs(z+ 1//2) < 1//10 ? imag(z) > 0 ? complex(pointseven,pointseven) : complex(pointseven,-pointseven) : z
+    else
+        if real(z) == convert(T,Inf)
+            k == 0 && return z
+            return z + complex(0,2*k*pi)
+        end
+        real(z) == -convert(T,Inf) && return -z + complex(0,(2*k+1)*pi)
+        w = log(z)
+        k != 0 ? w += complex(0, 2*k*pi) : nothing
+    end
+    _lambertw(z,w)
+end
+
+lambertw(z::Complex{T}, k::Integer) where T<:Integer = lambertw(float(z),k)
+
+# lambertw(e + 0im,k) is ok for all k
+#function lambertw(::Irrational{:e}, k::T) where T<:Integer
+function lambertw(::typeof(euler), k::T) where T<:Integer
+    k == 0 && return 1
+    @baddomain(k)
+end
+
+# Maybe this should return a float
+lambertw(::typeof(euler)) = 1
+#lambertw(::Irrational{:e}) = 1
+
+#lambertw{T<:Number}(x::T) = lambertw(x,0)
+lambertw(x::Number) = lambertw(x,0)
+
+lambertw(n::Irrational, args::Integer...) = lambertw(float(n),args...)
+
+### omega constant ###
+
+const omega_const_ = 0.567143290409783872999968662210355
+# The BigFloat `omega_const_bf_` is set via a literal in the function __init__ to prevent a segfault
+
+# maybe compute higher precision. converges very quickly
+function omega_const(::Type{BigFloat})
+  @compat  precision(BigFloat) <= 256 && return omega_const_bf_[]
```

Comment:
Weird indentation here, and why is `@compat` necessary?

Reply:
I think compatibility with v0.4.  There is already a deprecation in v0.5:
```
julia> get_bigfloat_precision()
WARNING: get_bigfloat_precision() is deprecated
```
In other words, this can be removed. (Although I am interested what you/community think is the proper indentation, were `@compat` necessary here.

Reply:
The Julia standard for indentation is 4 spaces. That's what's used in the Julia repository as well as the majority of packages.

We don't need compatibility with 0.4 or even 0.5 here, since the package requires 0.6 at a minimum.